### PR TITLE
Add gitea support

### LINKED ISSUE #743: Add gitea support
Since I'm mainly using gitea and not GitHub, i'd love to see gitea support in Knope.
This should be relatively simple, the documentation for it can be found here: https://try.gitea.io/api/swagger#/repository/repoCreatePullRequest

If there is interest, i would offer myself to implement it. 

Since this is a larger change, I will split this onto 2 PRs:
- [ ] refactor: Making references to github more neutral (using `forge` where we now use `github`)
- [ ] feature: Adding full gitea support

### PR DESCRIPTION
**Pull request by FallenValkyrie**
Took me a little longer, since i noticed that my initial idea would literally have required a full refactor, which was a little too large in scope. the above checklist is for keeping track of what will be added and this PR is here to track the progress and eventual questions.

closes #743

## Plans
- [x] Extending the configuration format and adding a `[gitea]` section
- [x] Allowing to generate the configuration from known gitea remotes
- [x] using gitea on the `CreatePullRequest` step
- [x] using gitea on the `Release` step
- [x] creating a new step `SelectGiteaIssue`
- [x] adding a new `forge` concept to the documentation to link forge integrations together in a convenient place
- [x] adding a new `gitea` section to the config file reference
- [x] multi-forge support as mentioned in https://github.com/knope-dev/knope/issues/743#issuecomment-1869778204

## Testing

- PR generated by knope on a copy of one of the projects i work on. (fixes will be pushed in a bit)
https://codeberg.org/FallenValkyrie/testing/pulls/5
- generated release
https://codeberg.org/FallenValkyrie/testing/releases/tag/v0.3.0
- Adding tests in `tests::gitea_release` and `tests::generate`
- selecting gitea issues (token has already been deleted, so no need to censor it)
An image showing knope successfully selecting gitea issues
- Using the same testing repo to see if knope correctly asks for the `GITEA_TOKEN`
An image of a Knope prepare-release workflow doing a lot of things and then asking for the Gitea token

## Notes
- Commits will be reorganized after review. This is simply because i don't feel like constantly reorganizing commits.
- The PR looks huge, but most of that is tests and test assets.
- Currently `ureq` (as far as i can tell) doesn't support sending multipart form data and the only library i could find, [`ureq_multipart`](https://docs.rs/ureq_multipart/latest/ureq_multipart/index.html), panics on IO errors and has messages in chinese (see [in its own codebase](https://gitee.com/eshangrao/ureq-multipart-toolkit/blob/master/src/lib.rs#L186-196)). Together i find those unacceptable design choices. issue: algesten/ureq#698

### CODE REVIEW

**Comment by FallenValkyrie** [comment]
The clippy lint that is triggering makes no sense. The error is a copy of `integrations::github::create_release::Error` with some minor changes. So why does it trigger here, but not there?

another one is the failing test. I do not know why it adds that unicode symbol there... i can't find that anywhere in the files.
```
---- version_go_mod stdout ----
thread 'version_go_mod' panicked at tests/gitea_release.rs:243:5:

---- expected: tests/gitea_release/version_go_mod/expected_go.mod
++++ actual:   In-memory
   1      - module codeberg.org/owner/repo // v1.1.0
        1 + module codeberg.org/owner/repo // v1.1.0∅
```
@dbanty would you know why this happens?

**Comment by FallenValkyrie** [comment]
> another one is the failing test. I do not know why it adds that unicode symbol there... i can't find that anywhere in the files.
> 
> ```
> ---- version_go_mod stdout ----
> thread 'version_go_mod' panicked at tests/gitea_release.rs:243:5:
> 
> ---- expected: tests/gitea_release/version_go_mod/expected_go.mod
> ++++ actual:   In-memory
>    1      - module codeberg.org/owner/repo // v1.1.0
>         1 + module codeberg.org/owner/repo // v1.1.0∅
> ```

Okay, this one makes a lot of sense now. The in-memory value uses `CRLF` to mark the end of the line where the file uses `LF`.

**Comment by FallenValkyrie** [comment]
https://codeberg.org/FallenValkyrie/testing/pulls/5

PR generated by knope on a copy of one of the projects i work on. (fixes will be pushed in a bit)

File: src/integrations/gitea/create_release.rs
Comment (FallenValkyrie) [inline src/integrations/gitea/create_release.rs:86]: Gitea requires a `multipart/form-data` encoded request here. ureq does not support sending `multipart/form-data` requests. see: https://github.com/algesten/ureq/issues/698

**Review by FallenValkyrie — COMMENTED** [review]
preliminary code check

File: .changeset/gitea_support.md
Comment (dbanty) [inline .changeset/gitea_support.md:25]:
```suggestion
but feel free to add more [here](https://github.com/knope-dev/knope/blob/main/src/config/toml/config.rs).
```
Can't point at an exact hash & line yet, but I think this link will be good enough

File: docs/src/content/docs/reference/Concepts/release.md
Comment (dbanty) [inline docs/src/content/docs/reference/Concepts/release.md:13]:
```suggestion
5. Optionally create a release (as part of the previous step) if [a forge is configured](/reference/concepts/forge)
```

File: docs/src/content/docs/reference/Config File/Steps/release.md
Comment (dbanty) [inline docs/src/content/docs/reference/Config File/Steps/release.md:6]:
```suggestion
If there is a [forge configured],
```

File: docs/src/content/docs/reference/Config File/Steps/release.md
Comment (dbanty) [inline docs/src/content/docs/reference/Config File/Steps/release.md:7]:
```suggestion
this creates a release with the same release notes that it added to the changelog (if any).
```

File: docs/src/content/docs/reference/Config File/Steps/release.md
Comment (dbanty) [inline docs/src/content/docs/reference/Config File/Steps/release.md:30]: Is there an equivalent of auto release notes for Gitea? We should clarify this sentence to either be generic or clarify that this only happens for GitHub/

File: src/config/toml/config.rs
Comment (dbanty) [inline src/config/toml/config.rs:84]:
```suggestion
    /// The domain of the Gitea instance
```

We're assuming https later, so I don't think an IP would work. Maybe that's also worth calling out in the docs for config.

File: src/integrations/gitea/create_release.rs
Comment (dbanty) [inline src/integrations/gitea/create_release.rs:58]: Oof. Maybe it's finally time to accept the fact that I'll need to pull in an async runtime. If we support pushing to multiple forges at once, there might finally be a parallelism benefit anyway.

Comment (dbanty) [reply src/integrations/gitea/create_release.rs:58]: Maybe we should error at the config level if both assets and `gitea` are configured, so someone doesn't find out post-release after digging through CI logs 😦

File: docs/src/content/docs/reference/Config File/Steps/create-pull-request.md
Comment (dbanty) [inline docs/src/content/docs/reference/Config File/Steps/create-pull-request.md:5]:
```suggestion
Create a pull request on every configured forge from the current branch to a specified branch. If a pull request for those already exists, this step will overwrite the title and body of the existing pull request.
```

or maybe copy the warning from the other page here?

File: src/step/issues/gitea.rs
Comment (dbanty) [inline src/step/issues/gitea.rs:36]:
```suggestion
                "Would prompt user to select an issue."
```

That state transition concept is a holdover from an older version—I probably need to clean that up in a few places still

File: src/step/mod.rs
Comment (dbanty) [inline src/step/mod.rs:46]:
```suggestion
    /// [`Issue::Selected`].
```

This is probably also wrong, I gotta look around the code base and figure out how to explain this better 😅

File: src/step/releases/mod.rs
Comment (dbanty) [inline src/step/releases/mod.rs:346]: Do you think I should set up a mirror of this repo on a Gitea instance and keep releases up to date there? Might help me catch issues if this integration ever gets out of date 🤔

File: tests/generate.rs
Comment (dbanty) [inline tests/generate.rs:142]:
```suggestion
/// Run `--generate` on a repo with a Codeberg remote.
```

File: tests/generate/gitea/knope.toml
Comment (dbanty) [inline tests/generate/gitea/knope.toml:28]: Should the host have the `https://`? I thought that was being added in code. The examples in the docs do not have a scheme.

File: tests/gitea_release.rs
Comment (dbanty) [inline tests/gitea_release.rs:35]: Though I do wonder if I should set up some integration tests to run only right before release.

File: tests/gitea_release/release_assets/dry_run_output.txt
Comment (dbanty) [inline tests/gitea_release/release_assets/dry_run_output.txt:1]: Think we should put the host here instead of `Gitea`?

**Review by dbanty — CHANGES_REQUESTED** [review]
Thanks so much for this 🤩! I left some questions & suggestions in the review. Here are a few more that didn't belong to any particular file:

1. There are probably more places where "GitHub" is mentioned that can be cleaned up to more generic forge lingo. I'll take a scan around on a later day.
2. Semi-related, it seems like Gitea Actions is supposed to be compatible with GitHub Actions, does that mean that the release recipes [preview release](https://knope.tech/recipes/preview-releases-with-pull-requests/) and [workflow dispatch](https://knope.tech/recipes/workflow-dispatch-releases/) will work as is? If so, it'd be nice to update them to mention that Gitea can be used too.
3. Do you think it's worth noting the relationship between Codeberg, Forgejo, and Gitea in the docs anywhere? Or is that relationship well known by their users (I had to look it up 😁)
4. Can you add a test that verifies if multiple forges are configured they are _all_ used?

Then, as a future follow-up, we probably should switch to a `[[forges]]` array to maintain the multi-forge capabilities but with more concise config. Historically I also include an auto-upgrade option with a deprecation so it's easy to update config. Adding the `host` option to GitHub isn't a bad idea either, for GitHub Enterprise.

File: .changeset/gitea_support.md
Comment (FallenValkyrie) [reply .changeset/gitea_support.md:25]: ah yes oops. i forgot that it sends me to my own fork xD

File: tests/generate.rs
Comment (FallenValkyrie) [reply tests/generate.rs:142]: oh no, the copy paste has been exposed xD

File: tests/generate/gitea/knope.toml
Comment (FallenValkyrie) [reply tests/generate/gitea/knope.toml:28]: yeah i forgot to update the doc examples. https is only assumed with known public instances. A private/internal instance however doesn't have to be https. Not everyone bothers using HTTPS within their own network after all. ^^

File: tests/gitea_release/release_assets/dry_run_output.txt
Comment (FallenValkyrie) [reply tests/gitea_release/release_assets/dry_run_output.txt:1]: mmmmh... yeah that makes sense to me

**Comment by FallenValkyrie** [comment]
> Do you think it's worth noting the relationship between Codeberg, Forgejo, and Gitea in the docs anywhere? Or is that relationship well known by their users (I had to look it up 😁)

mmmh... from my experience the people that use codeberg/forgejo know they are based on and fully compatible with gitea. But i think your idea to make this even more generic in the future might actually allow us to not have to. I'd say we mention Forgejo in the forge concept as a supported forge to be more clear. 

> Semi-related, it seems like Gitea Actions is supposed to be compatible with GitHub Actions, does that mean that the release recipes [preview release](https://knope.tech/recipes/preview-releases-with-pull-requests/) and [workflow dispatch](https://knope.tech/recipes/workflow-dispatch-releases/) will work as is? If so, it'd be nice to update them to mention that Gitea can be used too.

mmh I'd say we test if we can get an equivalent configuration in gitea working. Tho hosting my own CI instance already stretches my money to its max, so rn i can't do it. Since I'm a proponent of testing/verifying before mentioning things in docs I'd leave that to be added a later day. 

> Can you add a test that verifies if multiple forges are configured they are all used?

Sure can do ^^

> There are probably more places where "GitHub" is mentioned that can be cleaned up to more generic forge lingo. I'll take a scan around on a later day.

There were a few github specific examples that could stand cleaning. I've scanned the docs during my first try and it was quite a lot. Tho i would like to keep some of the GitHub specific stuff, since it mentions the CLI. 

> Then, as a future follow-up, we probably should switch to a [[forges]] array to maintain the multi-forge capabilities but with more concise config. Historically I also include an auto-upgrade option with a deprecation so it's easy to update config. Adding the host option to GitHub isn't a bad idea either, for GitHub Enterprise. 

I wanted to try this from the start, but it changed so much of the codebase that it would've been a pain to review. I do support doing that tho ^^

File: tests/gitea_release.rs
Comment (FallenValkyrie) [reply tests/gitea_release.rs:35]: mmmh... I think it would be enough to have a gitea instance (public or private) with a test repo.
The same would count for gh.

the way i did it is:
 I copied my own project (which was already using knope) into a test repo just to test knope releases, PRs and issues.

File: src/step/releases/mod.rs
Comment (FallenValkyrie) [reply src/step/releases/mod.rs:346]: I do use knope myself, so if it gets outdated i'll likely make an issue (and if i have time i'll also take care of implementation. 

I did see that you invited me
04F58095-396D-4818-B381-22D4B9063374

But testing the main repo is obviously always a good idea, so i support the idea ^^

File: src/config/toml/config.rs
Comment (FallenValkyrie) [reply src/config/toml/config.rs:84]: HTTPS is only assumed for known public instances ^^

File: src/integrations/gitea/create_release.rs
Comment (FallenValkyrie) [reply src/integrations/gitea/create_release.rs:58]: > Maybe we should error at the config level if both assets and `gitea` are configured, so someone doesn't find out post-release after digging through CI logs 😦

yeah, seems like a good idea to me, It's how i figured it out ^^

> Oof. Maybe it's finally time to accept the fact that I'll need to pull in an async runtime. If we support pushing to multiple forges at once, there might finally be a parallelism benefit anyway.

If you do decide to do so, i'd recommend the `reqwest` crate. been using it for pretty much everything that needs to request stuff.

**Comment by FallenValkyrie** [comment]
I just noticed that you merged this into a local branch. Since I accepted the invite, i would have access to that branch. My question is: how would you like me to continue? Would you like me to switch to the local branch within the repo or do it on the fork I'm already working on and only do future contributions here?

**Comment by dbanty** [comment]
I didn’t intentionally merge it anywhere 🧐. Maybe something automatic that GitHub did? You’re more than welcome to work on a branch in here or on your fork, whatever is easiest!

**Comment by FallenValkyrie** [comment]
Idk, see here:
An image showing that the PR branch was merged into a local branch of the same name

Anyway, in that case I'll finish this PR on the fork, since that is simpler on both ends, and any future PRs will happen on a local branch ^^

File: docs/src/content/docs/reference/Config File/Steps/release.md
Comment (FallenValkyrie) [reply docs/src/content/docs/reference/Config File/Steps/release.md:30]: The features are identical, besides what is mentioned in the notes up top. I likely just missed this ^^

**Comment by FallenValkyrie** [comment]
There we go, that should (hopefully) fix the conflict. Tho when rebasing on main, i didn't get a conflict at all...

**Comment by dbanty** [comment]
Enforced that Gitea + assets aren't both declared and opened #779 to track adding in support (either manually with `ureq` or switching to something else)

**Comment by FallenValkyrie** [comment]
> Enforced that Gitea + assets aren't both declared

I could swear i did that... but it seems my brain never finished that thought...

**Review by dbanty — APPROVED** [review]
I think it's ready! Thanks again for pulling this together! I've added some follow-up issues in #779, #780, and #781 but I don't think this is worth holding up for any of those.

**Review by dbanty — APPROVED** [review]

**Pull request merged**
